### ISSUE TITLE
Reconsider the purpose and implementation of allow_dynamic_queries

### ISSUE THREAD

**Issue by dblock** [github/graphql-client#128]
tl;dr I propose to remove https://github.com/github/graphql-client/blob/1f87b183de2ccc375f8dc74395d5a410f12f0122/lib/graphql/client.rb#L294

Coming from https://github.com/ashkan18/graphlient/issues/24.

Currently graphql-client forces you to declare static module variables in the name of security and optimization. It makes two good arguments in https://github.com/github/graphql-client/blob/master/guides/dynamic-query-error.md:

1) Parsing a query and validating a query on every request adds performance overhead. It also prevents validation errors from being discovered until request time, rather than when the query is parsed at startup.

2) Building runtime GraphQL query strings with user input may lead to security issues. Always using a static query along with variables is a best practice.

This means well, however it forces a certain implementation on the caller that is very un-ruby and cumbersome to manage. It forces the client to keep a static around forever whereas that might not be what the client wants. It requires a client to reload if the schema changed, for example or lead to hard to debug scenarios. 

We should be able to define a client, parse a query, then execute it separately, achieving both (1) and (2) without imposing this. I would vote to make `allow_dynamic_queries` true by default and better document the tradeoffs rather than impose a certain style even if it's for the greater good.

Finally, having gone through the experience of a first time Ruby GraphQL client call I was surprised that I can't  just make a query. Creating a module, parsing a query and executing a query are much more complicated compared to just making a query with all its values because this requires understanding of GraphQL parameters. Compare:

```ruby
Client = ...

InvoiceQuery = Client.parse <<-GRAPHQL
    query($id: Int) {
      invoice(id: :$id) {
        id
        fee_in_cents
      }
    }
GRAPHQL

Client.query InvoiceQuery, id: 42
```

to 

```ruby
Client = ///

Client.query <<-GRAPHQL
  invoice(id: 42) {
    id
    fee_in_cents 
  }
GRAPHQL
```

It gets much hairier with complex types or mutations. We want this library to become more friendly to beginners.

Finally, the apps that define such modules make HTTP calls on start. This is bad because they will keep crashing if they cannot access the remote GraphQL endpoint then, which is rarely desired.

**Comment by dblock** [comment]
Whoops wrong repo.

**Comment by egilburg** [comment]
I second the request to allow opt-out of static queries (checking for them by default is a good idea as a safety measure, but I need an escape valve).

The reason we can't use constant queries in our application is that **our schema itself is dynamically defined**. Our users can create or modify tables at runtime, and we need to query against a table the structure of which isn't known until query time.

**Comment by magicodeuk** [comment]
It's ridiculous to insist on using constants for this. That's not what constants are for! If you have to create the schema and then parse them, then they are not constant at all. You use Rubocop in the project so you should be well aware that this is bad practice.

**Comment by jeffgran** [comment]
This also seems to force me to hard-code the url of the graphql api, so I can't have it be configured in an initializer or anything.

```ruby
# must put query in a constant:
MyQuery = MyClient.parse(...)

# therefore must parse at compile time because 
"dynamic constant assignment (SyntaxError)"

# therefore must define http "execute" adapter at compile time while defining MyClient:
MyClient = GraphQL::Client.new(schema: schema, execute: GraphQL::Client::HTTP.new("http://hard-coded-url.com"))

#can't change the http executor at runtime:
"undefined method `execute=' for #<GraphQL::Client:0x00007fefc68b0138>"
```

What is the recommended way to test this code? I can't figure out a way to actually change the server at runtime to a test server... currently hacking with this:

```ruby
MyClient.instance_variable_set(:@execute, self.http)
```

geez...

**Comment by dblock** [comment]
@jeffgran I recommend you use VCR to record HTTP requests and responses. But if you need to make real requests during tests, do the dynamic thing. You'll probably find graphlient (built on top of this) friendlier, too.

**Comment by josh** [comment]
I still stand behind enforcing static queries for the scope of this project. It's a fundamental assumption around how response structures are preallocated, validated and potentially persisted to a server. 

If you want dynamic queries, consider using another library like [graphlient](https://github.com/ashkan18/graphlient).

**Comment by dblock** [comment]
@josh I am a co-author of Graphlient and it's built on top of this library, hence the discussion.

**Comment by josh** [comment]
> I don't love the idea of pulling this code apart into https://github.com/ashkan18/graphlient to remove the dependency.

I think where we can work together is trying to upstream common code into graphql-ruby itself. There's many common patterns around inspecting and processing GraphQL query ASTs that would benefit many use cases.

**Comment by dblock** [comment]
I really like that idea @josh. Care to open a new issue here and maybe lay out some practical first steps? I can try to help. What are the first parts you'd move?

**Comment by josh** [comment]
Hey @dblock, I support the idea of upstreaming parts to graphql-ruby. I don't have the bandwidth to push it forward personally. But I would gladly merge any PRs here that delete a bunch of code as it now exists in graphql-ruby. 🔥 is also good! Thanks for being patient.

**Comment by robinboening** [comment]
Did this discussion just stop here or is there a continuation somewhere else?
